{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cas protein and spacer cassette analysis \n",
    "### Relating expression to spacer cassette size (hopefully)\n",
    "\n",
    "Plan:\n",
    "1. Filter Cas protein results in database to only those that are near a CRISPR casette, and/or filter only high e-value hits or certain protein types. \n",
    "2. Get neighbouring casette lengths (copy code from locus number of spacers analysis)\n",
    "3. Get DNA for *cas* genes and get codons, run through a codon usage analyzer\n",
    "\n",
    "Will need to download genomes as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import numpy as np\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import mlab\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "from tqdm import tqdm\n",
    "\n",
    "sns.set_palette(\"husl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Url of the phageParser API\n",
    "apiurl = 'https://phageparser.herokuapp.com'\n",
    "#Get the initial page for listing of accessible objects and get url for organisms and casproteins\n",
    "url_json=requests.get(apiurl).json()\n",
    "organisms_url = url_json['organisms']\n",
    "cas_proteins_url = url_json['casproteins']\n",
    "organism_cas_url = url_json['organismcas']\n",
    "loci_url = url_json['loci']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Iterate through each page and merge the json response into a dictionary for cas proteins\n",
    "cas_protein_dict = {}\n",
    "r=requests.get(cas_proteins_url)\n",
    "last_page = r.json()['meta']['total_pages']\n",
    "for page in range(1,last_page+1):\n",
    "    url = cas_proteins_url+'?page={}'.format(page)\n",
    "    payload = requests.get(url).json()\n",
    "    cas_protein_objs = payload['cas_proteins']\n",
    "    for cas_protein_obj in cas_protein_objs:\n",
    "        cas_protein_dict[cas_protein_obj['id']] = cas_protein_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Iterate through each page and merge the json response into a dictionary for organisms\n",
    "organism_dict = {}\n",
    "r=requests.get(organisms_url)\n",
    "last_page = r.json()['meta']['total_pages']\n",
    "for page in range(1,last_page+1):\n",
    "    url = organisms_url+'?page={}&include[]=cas_proteins&include[]=loci.spacers'.format(page)\n",
    "    payload = requests.get(url).json()\n",
    "    organism_objs = payload['organisms']\n",
    "    for organism_obj in organism_objs:\n",
    "        organism_dict[organism_obj['id']] = organism_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/446 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total pages: 446\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 446/446 [09:23<00:00,  1.34s/it]\n"
     ]
    }
   ],
   "source": [
    "# Iterate through each page and merge the json response into a dictionary for organism_cas\n",
    "# this takes a long time\n",
    "organism_cas_dict = {}\n",
    "r = requests.get(organism_cas_url)\n",
    "last_page = r.json()['meta']['total_pages']\n",
    "print(\"total pages: \" + str(last_page))\n",
    "for page in tqdm(range(1, last_page+1)):\n",
    "    url = organism_cas_url+'?page={}'.format(page)\n",
    "    payload = requests.get(url).json()\n",
    "    organism_cas_objs = payload['organism_cas_proteins']\n",
    "    for organism_cas_obj in organism_cas_objs:\n",
    "        organism_cas_dict[organism_cas_obj['id']] = organism_cas_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through each page and merge the json response into a dictionary for loci\n",
    "loci_dict = {}\n",
    "r = requests.get(loci_url)\n",
    "last_page = r.json()['meta']['total_pages']\n",
    "print(\"total pages: \" + str(last_page))\n",
    "for page in tqdm(range(1, last_page+1)):\n",
    "    url = loci_url+'?page={}&include[]=spacers'.format(page)\n",
    "    payload = requests.get(url).json()\n",
    "    loci_objs = payload['locus']\n",
    "    for locus_obj in loci_objs:\n",
    "        loci_dict[locus_obj['id']] = locus_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 1,\n",
       " 'organism': 2170,\n",
       " 'genomic_start': 105255,\n",
       " 'genomic_end': 105651,\n",
       " 'spacers': [{'id': 1,\n",
       "   'sequence': 'AGCATGCCGATATTCCAATCTTTTGATTGGAGTT',\n",
       "   'length': 34},\n",
       "  {'id': 4, 'sequence': 'TTGTTCGCTCTATCGAGTTTGATGCTCGCCATAAG', 'length': 35},\n",
       "  {'id': 2, 'sequence': 'TACATTGGTAAGATTTTTAGACGAGGACATGTGTT', 'length': 35},\n",
       "  {'id': 3, 'sequence': 'TACAGACCTCCCTGCTTCTGAGCCGAATACAATCTTC', 'length': 37},\n",
       "  {'id': 5, 'sequence': 'CTCATAAATGCGAGGGGTCATGCAGCCGCCCGTA', 'length': 34}]}"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_dict[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dictionary with the info and positions of all cas proteins in each organism\n",
    "# note: this may be longer than the list of cas proteins in the organism dict, since some proteins may appear twice\n",
    "from collections import defaultdict\n",
    "\n",
    "organism_to_cas_dict = defaultdict(dict)\n",
    "for i, values in organism_cas_dict.items():\n",
    "    organism_to_cas_dict[values['organism']][i] = values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "533"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism_cas_dict[1]['organism']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retain only cas proteins near a cassette"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "distance_from_locus = 50000 # base pairs, look up what would be a good value\n",
    "# a cas9 sequence that I randomly found is like 9000 bp, so 50 000 seems like a reasonable cutoff.\n",
    "# also the cutoff doesn't really matter - the numbers are the same whether the cutoff is 50 000 or 500 000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism_names = []\n",
    "\n",
    "loci = []\n",
    "\n",
    "cas_genes = []\n",
    "\n",
    "cas_near_cassette_dict = defaultdict(dict)\n",
    "\n",
    "for organism_id in organism_dict.keys():\n",
    "\n",
    "    # search_window[0] must be less than search_window[1]\n",
    "\n",
    "    search_windows = {}\n",
    "    for locus in organism_dict[organism_id]['loci']:\n",
    "        # check that start is less than end\n",
    "        if not locus['genomic_start'] < locus['genomic_end']:\n",
    "            print('end is less than start')\n",
    "        search_windows[locus['id']] = {'window': [locus['genomic_start'] - 1000, locus['genomic_end'] + 1000]}\n",
    "    \n",
    "    for locus_id, windows in search_windows.items():\n",
    "        for cas_id, values in organism_to_cas_dict[organism_id].items():\n",
    "            #print(values['genomic_start'])\n",
    "            if values['genomic_start'] > windows['window'][0] and values['genomic_end'] < windows['window'][1]:\n",
    "                #print(organism_id, organism_dict[organism_id]['accession'], organism_dict[organism_id]['name'], locus_id)\n",
    "                organism_names.append(\"%s %s\\n\" %(organism_dict[organism_id]['accession'], organism_dict[organism_id]['name']))\n",
    "                loci.append(locus_id)\n",
    "                cas_genes.append(values['casprotein'])\n",
    "                cas_near_cassette_dict[organism_id][cas_id] = values\n",
    "                cas_near_cassette_dict[organism_id][cas_id]['locus'] = locus_id # also add locus id\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism_names = list(set(organism_names))\n",
    "\n",
    "with open('organism_names.txt', 'w') as file:\n",
    "    for line in organism_names:\n",
    "        file.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "473\n",
      "590\n",
      "202\n",
      "731\n",
      "394\n"
     ]
    }
   ],
   "source": [
    "print(len(set(organism_names)))\n",
    "print(len(set(loci)))\n",
    "print(len(set(cas_genes)))\n",
    "print(len(cas_genes))\n",
    "print(len(cas_protein_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download all organism genomes that were found to be hits\n",
    "!cat organism_names.txt | cut -d\" \" -f1 > organism_accessions.txt\n",
    "#!cat organism_accessions.txt | python ../parserscripts/acc2gb.py your@email.com nuccore fasta > cas_bacteria.dat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Need CDS regions for the AnaCoDa thing, so to download CDS fasta I think use `fasta_cds_na` (nucleotides for cds) or `fasta_cds_aa` (translated amino acids for cds). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1200252it [01:45, 11389.05it/s]\n"
     ]
    }
   ],
   "source": [
    "# split fasta file for all organisms into separated by organism\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "current_accession = 0\n",
    "past_accession = 0\n",
    "\n",
    "for record in tqdm(SeqIO.parse(\"cas_bacteria.dat\", \"fasta\")):\n",
    "    header = record.id\n",
    "    accession = header.rsplit('|')[1].rsplit('.')[0]\n",
    "    current_accession = accession\n",
    "    if current_accession != past_accession: # start a new file\n",
    "        with open(\"%s.fasta\" %accession, \"w\") as output_handle:\n",
    "            SeqIO.write(record, output_handle, \"fasta\")\n",
    "    else:\n",
    "        with open(\"%s.fasta\" %accession, \"a\") as output_handle:\n",
    "            SeqIO.write(record, output_handle, \"fasta\")\n",
    "    past_accession = current_accession\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "accession_to_organism_id = {}\n",
    "for key, values in organism_dict.items():\n",
    "    accession_to_organism_id[values['accession']] = key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to link cas genes to their position in the cds list - can be merged with loop above\n",
    "for record in tqdm(SeqIO.parse(\"cas_bacteria.dat\", \"fasta\")):\n",
    "    header = record.id\n",
    "    accession = header.rsplit('|')[1].rsplit('.')[0]\n",
    "    org_id = accession_to_organism_id[accession]\n",
    "    cds_number = int(header.split('_')[-1])\n",
    "    description = record.description\n",
    "    \n",
    "    for key, value in cas_near_cassette_dict[org_id].items():\n",
    "        try: \n",
    "            start = description.index(str(value['genomic_start']))\n",
    "            end = description.index(str(value['genomic_end']))\n",
    "            cas_near_cassette_dict[org_id][key]['cds_number'] = cds_number\n",
    "            #print(org_id)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis plan\n",
    "\n",
    "1. Get expression model data using the R package, save file as `accession-expression.csv`\n",
    "2. Load expression data using pandas\n",
    "3. Get cas gene ids: \n",
    "```\n",
    "cas_gene_positions = []\n",
    "for val in cas_near_cassette_dict[accession_to_organism_id['NZ_CP015184']].values():\n",
    "    cas_gene_positions.append(val['cds_number'])\n",
    "```\n",
    "4. The gene ids-1 are the index for the expression data frame (b/c zero indexing) **expression levels**\n",
    "5. Get associated locus size: \n",
    "```\n",
    "associated_loci = []\n",
    "loci_lengths = []\n",
    "for val in cas_near_cassette_dict[accession_to_organism_id['NZ_CP015184']].values():\n",
    "    associated_loci.append(val['locus'])\n",
    "    loci_lengths.append(len(loci_dict[val['locus']]['spacers'])\n",
    "```\n",
    "6. Multiple genes may be associated with the same locus, but now `cas_gene_positions` and `associated_loci` should be the same length to match them up (locus indices may be repeated). **locus lengths**\n",
    "\n",
    "Annddd.... done! Make a sweet plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "expression = pd.read_csv(\"NZ_CP015184-expression.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([267., 615., 530., 310., 171.,  96.,  52.,  32.,  17.,  16.,  15.,\n",
       "          5.,   8.,  12.,   6.,   4.,   1.,   1.,   1.,   1.]),\n",
       " array([0.19174138, 0.46599141, 0.74024145, 1.01449148, 1.28874152,\n",
       "        1.56299155, 1.83724159, 2.11149162, 2.38574166, 2.65999169,\n",
       "        2.93424173, 3.20849176, 3.4827418 , 3.75699183, 4.03124187,\n",
       "        4.3054919 , 4.57974194, 4.85399197, 5.12824201, 5.40249204,\n",
       "        5.67674208]),\n",
       " <a list of 20 Patch objects>)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAD5NJREFUeJzt3X+onuV9x/H3Z8Y2xf6I1qOEJCwdDV3HoOoOLkMom+mK2tLkj8pathokkH/csGSQ2v1TBvuj9Y/YCkMIplvcujqxLYZOXENUSmHanmiqtmlnJs4ckpnT+qN1khXb7/44V9az5JjznJzn5Dm58n7Bw33d133dz/neiJ/n8nru5zZVhSSpX78x6gIkSYvLoJekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+glqXMGvSR1btmoCwC49NJLa+3ataMuQ5LOKfv37/9JVY3NNW5JBP3atWuZmJgYdRmSdE5J8p+DjHPpRpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXOoJekzhn0ktQ5g16SOrckfhk7Sse33X7G5y7fsX2IlUjS4hhoRp9kRZL7k/woycEkf5DkkiR7kzzbthe3sUlyZ5JDSZ5KctXiXoIk6XQGXbr5EvBQVf028AHgIHAbsK+q1gH72j7A9cC69toK3DXUiiVJ8zJn0Cd5J/BBYBdAVf2iql4BNgK727DdwKbW3gjcU9MeA1YkWTn0yiVJAxlkRv9bwBTwd0meTHJ3kouAy6vqKEDbXtbGrwIOzzh/svVJkkZgkKBfBlwF3FVVVwL/za+XaWaTWfrqlEHJ1iQTSSampqYGKlaSNH+DBP0kMFlVj7f9+5kO/hdPLMm07bEZ49fMOH81cOTkN62qnVU1XlXjY2NzPjdfknSG5gz6qvov4HCS97WuDcAPgT3A5ta3GXigtfcAN7W7b9YDr55Y4pEknX2D3kf/F8BXkrwFeA64mekPifuSbAFeAG5sYx8EbgAOAa+3sZKkERko6KvqADA+y6ENs4wt4JYF1iVJGhIfgSBJnTPoJalzBr0kdc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXOoJekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+glqXOD/j9jNYvj224/43OX79g+xEok6c05o5ekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdGyjokzyf5OkkB5JMtL5LkuxN8mzbXtz6k+TOJIeSPJXkqsW8AEnS6c1nRv9HVXVFVY23/duAfVW1DtjX9gGuB9a111bgrmEVK0mav4Us3WwEdrf2bmDTjP57atpjwIokKxfwdyRJCzBo0BfwrST7k2xtfZdX1VGAtr2s9a8CDs84d7L1/T9JtiaZSDIxNTV1ZtVLkuY06CMQrqmqI0kuA/Ym+dFpxmaWvjqlo2onsBNgfHz8lOOSpOEYaEZfVUfa9hjwDeBq4MUTSzJte6wNnwTWzDh9NXBkWAVLkuZnzqBPclGSd5xoAx8GngH2AJvbsM3AA629B7ip3X2zHnj1xBKPJOnsG2Tp5nLgG0lOjP+nqnooyfeA+5JsAV4AbmzjHwRuAA4BrwM3D71qSdLA5gz6qnoO+MAs/T8FNszSX8AtQ6lOkrRg/jJWkjpn0EtS5wx6SeqcQS9JnTPoJalzBr0kdc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXOoJekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+glqXMGvSR1zqCXpM4Z9JLUuYGDPskFSZ5M8s22/54kjyd5Nsk/J3lL639r2z/Ujq9dnNIlSYOYz4z+VuDgjP0vAHdU1TrgZWBL698CvFxV7wXuaOMkSSMyUNAnWQ18BLi77Qe4Fri/DdkNbGrtjW2fdnxDGy9JGoFBZ/RfBLYDv2r77wZeqao32v4ksKq1VwGHAdrxV9t4SdIIzBn0ST4KHKuq/TO7ZxlaAxyb+b5bk0wkmZiamhqoWEnS/A0yo78G+FiS54F7mV6y+SKwIsmyNmY1cKS1J4E1AO34u4CXTn7TqtpZVeNVNT42Nragi5Akvbk5g76qPltVq6tqLfAJ4OGq+lPgEeDjbdhm4IHW3tP2accfrqpTZvSSpLNjIffRfwbYluQQ02vwu1r/LuDdrX8bcNvCSpQkLcSyuYf8WlU9Cjza2s8BV88y5jhw4xBqkyQNgb+MlaTOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXOoJekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+glqXMGvSR1zqCXpM4Z9JLUOYNekjpn0EtS5wx6SeqcQS9JnTPoJalzBr0kdc6gl6TOLZtrQJLlwLeBt7bx91fV55K8B7gXuAR4AvhUVf0iyVuBe4DfA34K/ElVPb9I9Z+zjm+7/YzPXb5j+xArkdS7QWb0/wNcW1UfAK4ArkuyHvgCcEdVrQNeBra08VuAl6vqvcAdbZwkaUTmDPqa9lrbvbC9CrgWuL/17wY2tfbGtk87viFJhlaxJGleBlqjT3JBkgPAMWAv8B/AK1X1RhsyCaxq7VXAYYB2/FXg3cMsWpI0uIGCvqp+WVVXAKuBq4H3zzasbWebvdfJHUm2JplIMjE1NTVovZKkeZrXXTdV9QrwKLAeWJHkxJe5q4EjrT0JrAFox98FvDTLe+2sqvGqGh8bGzuz6iVJc5oz6JOMJVnR2m8DPgQcBB4BPt6GbQYeaO09bZ92/OGqOmVGL0k6O+a8vRJYCexOcgHTHwz3VdU3k/wQuDfJ3wBPArva+F3APyQ5xPRM/hOLULckaUBzBn1VPQVcOUv/c0yv15/cfxy4cSjVSZIWzF/GSlLnDHpJ6pxBL0mdM+glqXMGvSR1zqCXpM4Nch/9kraQx/1K0vnAGb0kdc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXOoJekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+glqXMGvSR1zqCXpM7NGfRJ1iR5JMnBJD9IcmvrvyTJ3iTPtu3FrT9J7kxyKMlTSa5a7IuQJL25QWb0bwB/WVXvB9YDtyT5HeA2YF9VrQP2tX2A64F17bUVuGvoVUuSBjZn0FfV0ap6orV/DhwEVgEbgd1t2G5gU2tvBO6paY8BK5KsHHrlkqSBzGuNPsla4ErgceDyqjoK0x8GwGVt2Crg8IzTJlvfye+1NclEkompqan5Vy5JGsjAQZ/k7cDXgE9X1c9ON3SWvjqlo2pnVY1X1fjY2NigZUiS5mmgoE9yIdMh/5Wq+nrrfvHEkkzbHmv9k8CaGaevBo4Mp1xJ0nwNctdNgF3AwaraMePQHmBza28GHpjRf1O7+2Y98OqJJR5J0tm3bIAx1wCfAp5OcqD1/RXweeC+JFuAF4Ab27EHgRuAQ8DrwM1DrViSNC9zBn1VfYfZ190BNswyvoBbFliXJGlI/GWsJHXOoJekzhn0ktQ5g16SOjfIXTdaYo5vu/2Mz12+Y/sQK5F0LnBGL0mdM+glqXMGvSR1zqCXpM4Z9JLUOYNekjpn0EtS5wx6SeqcQS9JnTPoJalzBr0kdc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHVuzqBP8uUkx5I8M6PvkiR7kzzbthe3/iS5M8mhJE8luWoxi5ckzW2QGf3fA9ed1HcbsK+q1gH72j7A9cC69toK3DWcMiVJZ2rOoK+qbwMvndS9Edjd2ruBTTP676lpjwErkqwcVrGSpPk70zX6y6vqKEDbXtb6VwGHZ4ybbH2SpBEZ9pexmaWvZh2YbE0ykWRiampqyGVIkk5YdobnvZhkZVUdbUszx1r/JLBmxrjVwJHZ3qCqdgI7AcbHx2f9MNDwHd92+4LOX75j+5AqkXS2nOmMfg+wubU3Aw/M6L+p3X2zHnj1xBKPJGk05pzRJ/kq8IfApUkmgc8BnwfuS7IFeAG4sQ1/ELgBOAS8Dty8CDVLkuZhzqCvqk++yaENs4wt4JaFFiVJGh5/GStJnTPoJalzBr0kdc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXuTB9TrPPUQh5z7COOpdFwRi9JnTPoJalzBr0kdc6gl6TOGfSS1DmDXpI65+2VOmu8NVMaDWf0ktQ5g16SOmfQS1LnDHpJ6pxfxuqcsJAvchfCL4HVA4NeOo2FfsD4QaGlYFGCPsl1wJeAC4C7q+rzi/F3JM3OW1k109CDPskFwN8CfwxMAt9LsqeqfjjsvyX1bFTLVerPYszorwYOVdVzAEnuBTYCBr3OO+diWI+yZv9rYnEsRtCvAg7P2J8Efn8R/o6kzpyLH4wLdTY+3BYj6DNLX50yKNkKbG27ryX58UlDLgV+MuTaloqerw36vr6erw36vr6leW13fGYhZ//mIIMWI+gngTUz9lcDR04eVFU7gZ1v9iZJJqpqfPjljV7P1wZ9X1/P1wZ9X1/P1zaXxfjB1PeAdUnek+QtwCeAPYvwdyRJAxj6jL6q3kjy58C/Mn175Zer6gfD/juSpMEsyn30VfUg8OAC3+ZNl3U60PO1Qd/X1/O1Qd/X1/O1nVaqTvmeVJLUER9qJkmdW3JBn+S6JD9OcijJbaOuZ5iSfDnJsSTPjLqWYUuyJskjSQ4m+UGSW0dd0zAlWZ7ku0m+367vr0dd07AluSDJk0m+Oepahi3J80meTnIgycSo6znbltTSTXt8wr8z4/EJwCd7eXxCkg8CrwH3VNXvjrqeYUqyElhZVU8keQewH9jU0T+7ABdV1WtJLgS+A9xaVY+NuLShSbINGAfeWVUfHXU9w5TkeWC8qpbeffRnwVKb0f/f4xOq6hfAiccndKGqvg28NOo6FkNVHa2qJ1r758BBpn8l3YWa9lrbvbC9ls4saYGSrAY+Atw96lo0fEst6Gd7fEI3YXG+SLIWuBJ4fLSVDFdb2jgAHAP2VlVP1/dFYDvwq1EXskgK+FaS/e1X+eeVpRb0Az0+QUtXkrcDXwM+XVU/G3U9w1RVv6yqK5j+tffVSbpYfkvyUeBYVe0fdS2L6Jqqugq4HrilLaOeN5Za0A/0+AQtTW3t+mvAV6rq66OuZ7FU1SvAo8B1Iy5lWK4BPtbWse8Frk3yj6Mtabiq6kjbHgO+wfQy8XljqQW9j084R7UvK3cBB6tqx6jrGbYkY0lWtPbbgA8BPxptVcNRVZ+tqtVVtZbpf+cerqo/G3FZQ5PkonaDAEkuAj4MdHfn2+ksqaCvqjeAE49POAjc19PjE5J8Ffg34H1JJpNsGXVNQ3QN8CmmZ4MH2uuGURc1RCuBR5I8xfSEZG9VdXcbYqcuB76T5PvAd4F/qaqHRlzTWbWkbq+UJA3fkprRS5KGz6CXpM4Z9JLUOYNekjpn0EtS5wx6SeqcQS9JnTPoJalz/wu2LveRkOMVwAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(expression.PHI, bins = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism_id = 25\n",
    "\n",
    "search_windows = {}\n",
    "for locus in organism_dict[organism_id]['loci']:\n",
    "    # check that start is less than end\n",
    "    if not locus['genomic_start'] < locus['genomic_end']:\n",
    "        print('end is less than start')\n",
    "    search_windows[locus['id']] = {'window': [locus['genomic_start'] - 1000, locus['genomic_end'] + 1000]}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'id': 100592, 'organism': 25, 'casprotein': 305, 'genomic_start': 108876, 'genomic_end': 109115, 'evalue': 5.3e-09}\n",
      "{'id': 100593, 'organism': 25, 'casprotein': 140, 'genomic_start': 109222, 'genomic_end': 109647, 'evalue': 2.7e-74}\n",
      "{'id': 100593, 'organism': 25, 'casprotein': 140, 'genomic_start': 109222, 'genomic_end': 109647, 'evalue': 2.7e-74}\n"
     ]
    }
   ],
   "source": [
    "plot_regions = []\n",
    "for loci in organism_dict[organism_id]['loci']:\n",
    "    #print(loci['genomic_start'], loci['genomic_end'])\n",
    "    plot_regions.append({'position': [loci['genomic_start'], loci['genomic_end']], 'label': 'casette'})\n",
    "    \n",
    "for locus_id, windows in search_windows.items():\n",
    "    for cas_id, values in organism_to_cas_dict[organism_id].items():\n",
    "        #print(values['genomic_start'])\n",
    "        if values['genomic_start'] > windows['window'][0] and values['genomic_end'] < windows['window'][1]:\n",
    "            #print(str(organism_id) + 'yes')\n",
    "            plot_regions.append({'position': [values['genomic_start'], values['genomic_end']], \n",
    "                                 'label': cas_protein_dict[values['casprotein']]['gene']})\n",
    "            print(values)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'position': [120875, 121147], 'label': 'casette'},\n",
       " {'position': [109850, 110064], 'label': 'casette'},\n",
       " {'position': [115342, 115737], 'label': 'casette'},\n",
       " {'position': [110210, 110850], 'label': 'casette'},\n",
       " {'position': [108876, 109115], 'label': 'cse2gr11'},\n",
       " {'position': [109222, 109647], 'label': 'cas7'},\n",
       " {'position': [109222, 109647], 'label': 'cas7'}]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the NCBI database for this organism, the closest CDS regions for the cas genes are:\n",
    "\n",
    "```\n",
    "complement(108876..109115)\n",
    "                     /locus_tag=\"GZL_RS47195\"\n",
    "                     /inference=\"COORDINATES: similar to AA\n",
    "                     sequence:RefSeq:WP_006272258.1\"\n",
    "                     /note=\"Derived by automated computational analysis using\n",
    "                     gene prediction method: Protein Homology.\"\n",
    "                     /codon_start=1\n",
    "                     /transl_table=11\n",
    "                     /product=\"hypothetical protein\"\n",
    "                     /protein_id=\"WP_078876378.1\"\n",
    "                     /translation=\"MVLDPFGGTGTTAAAAKVLGRVGISVDLSPAYTRTAASPSPTPS\n",
    "                     ANGVARPAGEWGSSSPQSSNWSTLWWMARLRLTLS\"\n",
    "```\n",
    "\n",
    "```\n",
    "109222..109647\n",
    "                     /locus_tag=\"GZL_RS43525\"\n",
    "                     /old_locus_tag=\"GZL_p00118\"\n",
    "                     /inference=\"COORDINATES: ab initio prediction:GeneMarkS+\"\n",
    "                     /note=\"Derived by automated computational analysis using\n",
    "                     gene prediction method: GeneMarkS+.\"\n",
    "                     /codon_start=1\n",
    "                     /transl_table=11\n",
    "                     /product=\"hypothetical protein\"\n",
    "                     /protein_id=\"WP_040108171.1\"\n",
    "                     /translation=\"MHRDVTARPWAREREQARAWQAVVAARAEAAAARPEENVDVSVV\n",
    "                     APAPVVLPAPRPSAAATEPELEDADDDQDLVLEDLTREQVLGWRDRAATDHQIVHDHI\n",
    "                     ARYGEHSAQRLFTRAFVATVQRLSHLGHLDLGYTPWRQA\"\n",
    "```\n",
    "\n",
    "So the good news is, the windows that the database outputs correspond to true CDS regions in NCBI, and we can get the translation from NCBI if necessary. Perhaps we should download the genbank files? or will it work to translate them ourselves?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x7ff58ecdbb00>"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIsAAACPCAYAAAB6UVtrAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAGiVJREFUeJzt3X2QVeWd4PHvLy1JR0NAabUSWqdRcCOOBk2rVJGNYzC8TAyKo6PEiegmBZvSrIHMKhY7iTHJVkYxcceYREWijC/oWKXLRie+RCKbTIw0CRuCYmyQkVZqgqAQVGghv/2jDzeXppHb9O3uq3w/Vaf6nuftPufy9HO4v37OOZGZSJIkSZIkSQDv6e8OSJIkSZIkqXYYLJIkSZIkSVKJwSJJkiRJkiSVGCySJEmSJElSicEiSZIkSZIklRgskiRJkiRJUonBIkmSJEmSJJUYLJIkSZIkSVKJwSJJkiRJkiSVHNDfHeisoaEhm5qa+rsbkiRJkiRJ7xpLly59JTMPraRszQWLmpqaaGlp6e9uSJIkSZIkvWtExL9XWnavl6FFxLyI+ENE/G4P+RER/xQRrRHx24g4qSxvakQ8X2xTK+2UJEmSJEmS+kcl9yy6HZjwNvkTgRHFNg34AUBEHAJ8DTgVOAX4WkQc3JPOSpIkSZIkqXftNViUmYuBjW9T5CxgfnZ4ChgcER8CxgOPZebGzHwVeIy3DzpJkiRJkiSpn1XjnkVDgbVl+21F2p7SdxMR0+hYlcSRRx5ZhS7Vhq0zr61qe/XfuWKvZSYvPGmvZfrKA5N+3eM2PvhB+OMfe9bGwIGweXPfvmet6e5nIEmqTf19jvJ8IknqbW+99RZtbW1s3bq1W/VefBEye6dPEfBOClXU19fT2NjIgAED9rmNagSLoou0fJv03RMzbwFuAWhubu6lf169E1XjP8TdbePdFiiCd+cxSdL+qL/n8/5+f0nSu19bWxsDBw6kqamJiK7CCl17/fVe7BRw7LG92361ZCYbNmygra2NYcOG7XM7ldyzaG/agCPK9huBl98mXZIkSZIkaTdbt25lyJAh3QoU6c8igiFDhnR7ZVZn1QgWLQQuKp6KNhrYlJnrgEeAcRFxcHFj63FFmiRJkiRJUpcMFPVMNT6/vV6GFhH3AH8FNEREGx1POBsAkJk/BB4G/hpoBd4ALinyNkbEN4AlRVPXZObb3ShbkiRJkiRJ/WyvwaLMnLKX/AQu3UPePGDevnVNkiRJkiTtzyp9cNRfVtje7z679wdH9cSyZcv44he/yObNm6mrq2P27Nmcf/753Wpjw4YNnHvuuSxZsoSLL76Y733ve6W82bNnM3/+fF599VW2bNlS7e6XVOMyNEmSJEmSpP3egQceyPz581mxYgU/+clP+PKXv8xrr71Wcf3t27dTX1/PN77xDebMmbNb/mc+8xmefvrpana5S9V4GpokSZIkSdK7wkMPzefOO+cQEQwffgKf+MQkLr7469TV1TFo0CAWL17Mjh07mDVrFj/72c/Ytm0bl156KdOnT+eYY44ptfPhD3+Yww47jPXr1zN48GAefvhhZs6cSUNDAyeddBKrV6/mxz/+MVdffTUvv/wya9asoaGhgbvvvpuPf/zjtLa27ta30aNH98lnYLBIkiRJkiQJWLVqBfPmfYvbbvsFgwc3sGnTRqZPP40nn3yEoUOHllYJ3XbbbQwaNIglS5awbds2xowZw7hx43Z5XP3TTz9Ne3s7Rx99NFu3bmX69OksXryYYcOGMWXKrnf8Wbp0KT//+c95//vf36fHuydehiZJkiRJkgS0tDzB2LHnMnhwAwCDBh3CRz86hosvvphbb72VHTt2APDoo48yf/58Ro0axamnnsqGDRt4/vnnS+2sW7eOz33uc/zoRz/iPe95DytXruSoo44qBZM6B4smTZpUM4EicGWRJEmSJEkSAJm526Pnr7rqh+zY8SseeughRo0axbJly8hMbrzxRsaPH79bG5s3b+bTn/403/zmN0uXjXU8G2zPDjrooOodRBW4skiSJEmSJAk4+eSxPP74fbz22gYANm3aSFvbKk499VSuueYaGhoaWLt2LePHj+cHP/gBb731FgC///3vef3112lvb2fy5MlcdNFFnHfeeaV2P/KRj7B69WrWrFkDwL333tvnx9YdriySJEmSJEk1qf47lT3qvqWlOu939NHHcckls5k+/TTq6uo45pgTef31zfzDPzxPZjJ27Fg++tGPcsIJJ7BmzRpOOukkMpNDDz2UBx98kAceeIDFixezYcMGbr/9dgBuv/12Ro0axfe//30mTJhAQ0MDp5xyytv2o6mpic2bN9Pe3s6DDz7Io48+ysiRI7niiiu4++67eeONN2hsbOQLX/gCV199dXUOvkzsbSlUX2tubs6Wav0r97OtM6+tanuV/JJMXnhSVd+zJx6Y9Oset9Fp9d8+684wr9Z71poa+1WXJO2DWjhHeT6RJPWmZ599lmOPPbbb9Xo7jNDc3PM2tmzZwgc+8AEyk0svvZQRI0YwY8aMnjfcha4+x4hYmpkVHYmXoUmSJEmSJPWyW2+9lVGjRnHcccexadMmpk+f3t9d2iMvQ5MkSZIkSeplM2bM6LWVRNXmyiJJkiRJkiSVGCySJEmSJElSicEiSZIkSZIklRgskiRJkiRJUok3uJYkSZIkSTXp8Zs/VtX2Bn9saVXb62zRokW73MR65cqVLFiwgLPPPrtX37faKgoWRcQE4H8BdcDczPx2p/zvAqcXuwcCh2Xm4CJvB7C8yHsxMydVo+OSJEmSJEm15PTTT2fZsmUAbNy4keHDhzNu3Lh+7lX37TVYFBF1wE3Ap4A2YElELMzMZ3aWycwZZeW/BJxY1sSbmTmqel2WJEmSJEnqHQ89NJ8775xDRDB8+Amcccbfcuml36S9vZ0hQ4Zw1113cfjhh/Pkk09y+eWXAxARLF68mIEDB5bauf/++5k4cSIHHnhgfx3KPqtkZdEpQGtmrgaIiAXAWcAzeyg/BfhadbonSZIkSZLUN1atWsG8ed/ittt+weDBDWzatJGIYMaMp4gI5s6dy7XXXsv111/PnDlzuOmmmxgzZgxbtmyhvr5+l7YWLFjAzJkz++lIeqaSYNFQYG3ZfhtwalcFI+IvgGHAE2XJ9RHRAmwHvp2ZD+5jXyVJkiRJknpNS8sTjB17LoMHNwAwaNAhtLYuZ/z481m3bh3t7e0MGzYMgDFjxjBz5kwuvPBCzjnnHBobG0vtrFu3juXLlzN+/Ph+OY6equRpaNFFWu6h7AXA/Zm5oyztyMxsBj4L3BARR+/2BhHTIqIlIlrWr19fQZckSZIkSZKqKzOJ2DUMct11X+Kyyy5j+fLl3HzzzWzduhWAWbNmMXfuXN58801Gjx7NypUrS3Xuu+8+Jk+ezIABA/q0/9VSSbCoDTiibL8ReHkPZS8A7ilPyMyXi5+rgZ+x6/2Mdpa5JTObM7P50EMPraBLkiRJkiRJ1XXyyWN5/PH7eO21DQBs2rSRLVs2MXToUADuuOOOUtlVq1Zx/PHHc+WVV9Lc3LxLsOiee+5hypQpfdv5KqrkMrQlwIiIGAa8REdA6LOdC0XEfwIOBn5ZlnYw8EZmbouIBmAMcG01Oi5JkiRJkt7dzphe2aPuW1qq835HH30cl1wym+nTT6Ouro5jjjmRadOu5rzzzmPo0KGMHj2aF154AYAbbriBRYsWUVdXx8iRI5k4cSIAa9asYe3atZx22mnV6VQ/2GuwKDO3R8RlwCNAHTAvM1dExDVAS2YuLIpOARZkZvklascCN0fEn+hYxfTt8qeoSZIkSZIk1ZIzz5zKmWdO3SXtK185a7dyN954Y5f1m5qaeOmll3qlb32lkpVFZObDwMOd0r7aaf/qLur9G3B8D/onSZIkSZKkPlTJPYskSZIkSZK0nzBYJEmSJEmSpBKDRZIkSZIkSSoxWCRJkiRJkqQSg0WSJEmSJEkqqehpaJIkSZIkSX1t8sKTqtre7A//uqrtdbZo0SJmzJhR2l+5ciULFizg7LPP7tX3rTaDRZIkSZIkSVVw+umns2zZMgA2btzI8OHDGTduXD/3qvsMFkmSJEmSJBUeemg+d945h4hg+PATOOOMv+XSS79Je3s7Q4YM4a677uLwww/nySef5PLLLwcgIli8eDEDBw4stXP//fczceJEDjzwwP46lH1msEiSJEmSJAlYtWoF8+Z9i9tu+wWDBzewadNGIoIZM54iIpg7dy7XXnst119/PXPmzOGmm25izJgxbNmyhfr6+l3aWrBgATNnzuynI+kZg0WSJEmSJElAS8sTjB17LoMHNwAwaNAhtLYuZ/z481m3bh3t7e0MGzYMgDFjxjBz5kwuvPBCzjnnHBobG0vtrFu3juXLlzN+/Ph+OY6e8mlokiRJkiRJQGYSEbukXXfdl7jssstYvnw5N998M1u3bgVg1qxZzJ07lzfffJPRo0ezcuXKUp377ruPyZMnM2DAgD7tf7UYLJIkSZIkSQJOPnksjz9+H6+9tgGATZs2smXLJoYOHQrAHXfcUSq7atUqjj/+eK688kqam5t3CRbdc889TJkypW87X0VehiZJkiRJkmrSA5Mqe9R9S0t13u/oo4/jkktmM336adTV1XHMMScybdrVnHfeeQwdOpTRo0fzwgsvAHDDDTewaNEi6urqGDlyJBMnTgRgzZo1rF27ltNOO606neoHBoskSZIkSZIKZ545lTPPnLpL2le+ctZu5W688cYu6zc1NfHSSy/1St/6ipehSZIkSZIkqcRgkSRJkiRJkkoqChZFxISIeC4iWiNiVhf5F0fE+ohYVmxfKMubGhHPF9vUznUlSZIkSZJ2ysz+7sI7WjU+v73esygi6oCbgE8BbcCSiFiYmc90KnpvZl7Wqe4hwNeAZiCBpUXdV3vcc0mSJEmS9K5SX1/Phg0bGDJkyG6PsNfeZSYbNmygvr6+R+1UcoPrU4DWzFwNEBELgLOAzsGirowHHsvMjUXdx4AJwD371l1JkiRJkvRu1djYSFtbG+vXr+9WvVde6aUOFZ59tnfbr6b6+noaGxt71EYlwaKhwNqy/Tbg1C7K/U1EfAL4PTAjM9fuoe7QzhUjYhowDeDII4+srOeSJEmSJOldZcCAAQwbNqzb9UaO7IXOlNnfroyr5J5FXa376vwx/R+gKTNPAB4H7uhGXTLzlsxszszmQw89tIIuSZIkSZIkqTdUEixqA44o228EXi4vkJkbMnNbsXsr8LFK60qSJEmSJKl2VBIsWgKMiIhhEfFe4AJgYXmBiPhQ2e4kYOfVfI8A4yLi4Ig4GBhXpEmSJEmSJKkG7fWeRZm5PSIuoyPIUwfMy8wVEXEN0JKZC4H/FhGTgO3ARuDiou7GiPgGHQEngGt23uxakiRJkiRJtaeSG1yTmQ8DD3dK+2rZ66uAq/ZQdx4wrwd9lCRJkiRJUh+p5DI0SZIkSZIk7ScMFkmSJEmSJKnEYJEkSZIkSZJKDBZJkiRJkiSpxGCRJEmSJEmSSgwWSZIkSZIkqcRgkSRJkiRJkkoMFkmSJEmSJKnEYJEkSZIkSZJKDBZJkiRJkiSpxGCRJEmSJEmSSgwWSZIkSZIkqcRgkSRJkiRJkkoMFkmSJEmSJKnEYJEkSZIkSZJKDBZJkiRJkiSppKJgUURMiIjnIqI1ImZ1kT8zIp6JiN9GxE8j4i/K8nZExLJiW1jNzkuSJEmSJKm6DthbgYioA24CPgW0AUsiYmFmPlNW7DdAc2a+ERFfBK4Fzi/y3szMUVXutyRJkiRJknpBJSuLTgFaM3N1ZrYDC4Czygtk5qLMfKPYfQporG43JUmSJEmS1BcqCRYNBdaW7bcVaXvyeeBfy/brI6IlIp6KiLO7qhAR04oyLevXr6+gS5IkSZIkSeoNe70MDYgu0rLLghF/BzQDp5UlH5mZL0fEUcATEbE8M1ft0ljmLcAtAM3NzV22LUmSJEmSpN5XycqiNuCIsv1G4OXOhSLiDGA2MCkzt+1Mz8yXi5+rgZ8BJ/agv5IkSZIkSepFlQSLlgAjImJYRLwXuADY5almEXEicDMdgaI/lKUfHBHvK143AGOA8htjS5IkSZIkqYbs9TK0zNweEZcBjwB1wLzMXBER1wAtmbkQuA74APAvEQHwYmZOAo4Fbo6IP9ERmPp2p6eoSZIkSZIkqYZUcs8iMvNh4OFOaV8te33GHur9G3B8TzooSZIkSZKkvlPJZWiSJEmSJEnaTxgskiRJkiRJUonBIkmSJEmSJJUYLJIkSZIkSVKJwSJJkiRJkiSVGCySJEmSJElSicEiSZIkSZIklRgskiRJkiRJUonBIkmSJEmSJJUYLJIkSZIkSVKJwSJJkiRJkiSVGCySJEmSJElSicEiSZIkSZIklRgskiRJkiRJUonBIkmSJEmSJJVUFCyKiAkR8VxEtEbErC7y3xcR9xb5v4qIprK8q4r05yJifPW6LkmSJEmSpGrba7AoIuqAm4CJwEhgSkSM7FTs88CrmTkc+C7wj0XdkcAFwHHABOD7RXuSJEmSJEmqQZWsLDoFaM3M1ZnZDiwAzupU5izgjuL1/cDYiIgifUFmbsvMF4DWoj1JkiRJkiTVoEqCRUOBtWX7bUVal2UyczuwCRhSYV1JkiRJkiTViAMqKBNdpGWFZSqpS0RMA6YBHHnkkRV06Z2h/jtX9Pl7PjDp133+nr1p4ED44x973kZfv2et6e5nIEmqTf19jvJ8IkmqVb15jtwfz3+VBIvagCPK9huBl/dQpi0iDgAGARsrrEtm3gLcAtDc3LxbMEn7r82b94/3lCSpEp6jJEnqmufI6qrkMrQlwIiIGBYR76XjhtULO5VZCEwtXp8LPJGZWaRfUDwtbRgwAni6Ol2XJEmSJElSte11ZVFmbo+Iy4BHgDpgXmauiIhrgJbMXAjcBvxzRLTSsaLogqLuioi4D3gG2A5cmpk7eulYJEmSJEmS1EPRsQCodjQ3N2dLS0t/d0OSJEmSJOldIyKWZmZzRWVrLVgUEeuBf+/vfugdqwF4pb87IZVxTKoWOS5VaxyTqkWOS9Uax6R66i8y89BKCtZcsEjqiYhoqTRSKvUFx6RqkeNStcYxqVrkuFStcUyqL1Vyg2tJkiRJkiTtJwwWSZIkSZIkqcRgkd5tbunvDkidOCZVixyXqjWOSdUix6VqjWNSfcZ7FkmSJEmSJKnElUWSJEmSJEkqMVikmhAR8yLiDxHxu7K08yJiRUT8KSKaO5W/KiJaI+K5iBhflj6hSGuNiFll6cMi4lcR8XxE3BsR7y3S31fstxb5Tb1/tHqn6M64jIhPRcTSiFhe/PxkWd7HivTWiPiniIgi/ZCIeKwYl49FxMFFehTlWiPitxFxUl8et2pXd+fKIv/IiNgSEX9fluZcqarYh/P3CRHxyyJ/eUTUF+nOk6qabp6/B0TEHcX4ezYirirLc65UVexhTF4XESuLOeyBiBhclud3HfU7g0WqFbcDEzql/Q44B1hcnhgRI4ELgOOKOt+PiLqIqANuAiYCI4EpRVmAfwS+m5kjgFeBzxfpnwdezczhwHeLctJOt1PhuAReAT6TmccDU4F/Lsv7ATANGFFsO9ucBfy0GJc/LfahYwzvLDutqC9B98bkTt8F/nXnjnOlqux2Kj9/HwDcCfzXzDwO+CvgrSLbeVLVdDuVz5XnAe8rzt8fA6ZHRJNzparsdnYfk48Bf5mZJwC/B64Cv+uodhgsUk3IzMXAxk5pz2bmc10UPwtYkJnbMvMFoBU4pdhaM3N1ZrYDC4Czir9OfhK4v6h/B3B2WVt3FK/vB8bu/Gum1J1xmZm/ycyXi90VQH3x15wPAR/MzF9mx03i5tP1+Os8Ludnh6eAwUU72s91c64kIs4GVtMxJndyrlTVdHNMjgN+m5n/ryi3ITN3OE+q2ro5LhM4qAhmvh9oBzbjXKkq2sOYfDQztxe7TwGNxWu/66gmGCzSO9FQYG3ZfluRtqf0IcBrZZPxzvRd2iryNxXlpZ74G+A3mbmNjjHWVpZXPv4Oz8x1AMXPw4r0PY1lqWIRcRBwJfD1TlnOleovxwAZEY9ExK8j4ooi3XlS/el+4HVgHfAiMCczN+Jcqb71X/jzKmC/66gmHNDfHZD2QVfR8KTr4Ge+Tfm3a0vaJxFxHB1LfMftTOqi2N7GmONS1fB1Opakb+n0R8Q9jS/nSvW2A4CPAycDbwA/jYildKzi6Mx5Un3lFGAH8GHgYOD/RsTjOFeqj0TEbGA7cNfOpC6K+V1Hfc5gkd6J2oAjyvYbgZ2X/3SV/gody9MPKCLq5eV3ttVWLD8eRKclolKlIqIReAC4KDNXFclt/HlZMew6/v4jIj6UmeuKyyf+UFZnT2NcqtSpwLkRcS0wGPhTRGwFluJcqf7RBjyZma8ARMTDwEl03MfIeVL95bPATzLzLeAPEfELoJmO1RjOlepVETEVOBMYW1yGC37XUY3wMjS9Ey0ELijuBzOMjptbPg0sAUYUTwN4Lx03hltYTLyLgHOL+lOB/13W1tTi9bnAE2UTtVSx4gkWDwFXZeYvdqYXl038MSJGF9eIX0TX46/zuLwoOowGNu28DEOqVGb+58xsyswm4Abgf2bm93CuVP95BDghIg4svrScBjzjPKl+9iLwyWIsHQSMBlbiXKleFhET6LhcfFJmvlGW5Xcd1YbMdHPr9w24h45rxd+iIwL+eWBy8Xob8B/AI2XlZwOrgOeAiWXpf03H0wRWAbPL0o+iY5JtBf6FjqdeANQX+61F/lH9/Vm41c7WnXEJ/A867nmwrGw7rMhrpuMpLKuA7wFRpA+h4+k+zxc/DynSg46nXawClgPN/f1ZuNXG1t25sqze1cDfl+07V7pVZduH8/ff0XHD9d8B15alO0+6VW3r5vn7A8X8tgJ4BvjvZe04V7pVZdvDmGylYwXbzv83/rCsvN913Pp923kiliRJkiRJkrwMTZIkSZIkSX9msEiSJEmSJEklBoskSZIkSZJUYrBIkiRJkiRJJQaLJEmSJEmSVGKwSJIkSZIkSSUGiyRJkiRJklRisEiSJEmSJEkl/x9Xd7ePqJkCaQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1440x144 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize = (20,2))\n",
    "\n",
    "# get max range needed\n",
    "min_value = 10**10\n",
    "max_value = 0\n",
    "for region in plot_regions:\n",
    "    if region['position'][0] < min_value:\n",
    "        min_value = region['position'][0]\n",
    "    if region['position'][1] > max_value:\n",
    "        max_value = region['position'][1]\n",
    "     \n",
    "    #if region['label'] == 'casette':\n",
    "    #    colour = 'b'\n",
    "    #else:\n",
    "    #    colour = 'g'\n",
    "    if region['label'] == 'casette':\n",
    "        ax.fill_between([region['position'][0], region['position'][1]],0,1, color = 'b')\n",
    "    else:\n",
    "        ax.fill_between([region['position'][0], region['position'][1]],0,1, label = region['label'])\n",
    "\n",
    "#ax.set_xlim(1893000, 1910000)\n",
    "#ax.set_xlim(748000, 753000)\n",
    "ax.legend(loc='best')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NZ_CP003988\n",
      "Streptomyces sp. 769\n",
      "Fetching 1 entries from GenBank: NZ_CP003988\n"
     ]
    }
   ],
   "source": [
    "# download a bacterial genome\n",
    "\n",
    "accession = organism_dict[organism_id]['accession']\n",
    "print(accession)\n",
    "print(organism_dict[organism_id]['name'])\n",
    "!echo \"$accession\" | python ../parserscripts/acc2gb.py madeleine.bonsma@gmail.com nuccore fasta > \"$accession\".dat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "for record in SeqIO.parse(\"%s.dat\" %accession, \"fasta\"):\n",
    "    genome = record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Seq\n",
    "print(Seq.transcribe(Seq.reverse_complement(genome[109222:109647])))\n",
    "print(Seq.translate((genome[109222:109647])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "141.66666666666666"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genome[109222:109647])/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGCACCGGGACGTGACGGCCAGGCCGTGGGCGCGCGAGCGCGAACAGGCCCGCGCCTGGCAGGCGGTCGTCGCAGCGCGCGCCGAAGCCGCGGCCGCGCGGCCCGAGGAGAACGTCGACGTCTCGGTCGTTGCGCCGGCACCGGTCGTCTTGCCCGCGCCCCGCCCGTCCGCCGCCGCTACGGAGCCGGAGCTCGAGGACGCCGACGACGACCAGGATCTGGTCCTCGAGGACTTGACGCGCGAGCAGGTGCTGGGCTGGCGCGACCGCGCTGCCACCGATCACCAGATCGTCCACGACCACATCGCCCGGTACGGCGAGCACTCGGCGCAGAGGCTGTTCACCCGGGCATTCGTCGCGACCGTGCAGCGCCTGTCCCACCTCGGGCACCTCGACCTTGGCTACACCCCATGGAGGCAGGCGTGA\n"
     ]
    }
   ],
   "source": [
    "print(genome[109222:109647])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Current issue: these don't look like obvious proteins in the sense of having a start codon at one end and a stop at the other. Translate all ORFs? How to identify which are the actual codons? Maybe the hmmer documentation has some info about this."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:phageParser]",
   "language": "python",
   "name": "conda-env-phageParser-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
